### ISSUE TITLE
Dart should allow server-initiated TLS renegotiation

### ISSUE THREAD

**Issue by brianquinlan** [dart-lang/sdk#47841]
```
dart --version
Dart SDK version: 2.14.2 (stable) (Wed Sep 15 12:32:06 2021 +0200) on "macos_x64"
```

The original context of this issue is here: https://github.com/flutter/flutter/issues/41573#issuecomment-978723279

The TL;DR is that Dart rejects server-initiated TLS renegotiation.

Sample traffic:
![image](https://user-images.githubusercontent.com/8851008/144345433-f17eaf75-4cba-47f2-b96c-c0df09ff8f3c.png)

Dart is rejecting the attempt by the server to renegotiate. That is the default for BoringSSL:
https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#SSL_set_renegotiate_mode

If I modify dart to allow renegotiation e.g.
```
+  SSL_set_renegotiate_mode(ssl_, ssl_renegotiate_freely);
```

Then the output is the same as with `curl` (using both sockets directly and `HttpClient`) and with python. 

I'm not sure if enabling renegotiation is safe or not - RPC 5746 lays out some possible renegotiation attacks:
https://datatracker.ietf.org/doc/html/rfc5746#section-1

We could allow the user to control whether TLS renegotiation is allowed but this is a difficult issue to diagnose.

**Comment by brianquinlan** [comment]
Chromium allows renegotiation so maybe it is OK?
https://chromium.googlesource.com/chromium/src/+/lkgr/net/socket/ssl_client_socket_impl.cc#912

**Comment by brianquinlan** [comment]
It looks like the reason for the poor diagnostics is that we don't check the result of `SSL_read`:
https://github.com/dart-lang/sdk/blob/main/runtime/bin/secure_socket_filter.cc

**Comment by Cubxity** [comment]
I just ran across the same issue with Xbox Live's API (`https://user.auth.xboxlive.com/user/authenticate`). 

**Code:**

```dart
final payload = {
  "Properties": {
    "AuthMethod": "RPS",
    "SiteName": "user.auth.xboxlive.com",
    "RpsTicket": "d=$accessToken"
  },
  "RelyingParty": "http://auth.xboxlive.com",
  "TokenType": "JWT"
};

final body = jsonEncode(payload);

final res = await _client.post(
  Uri.parse("https://user.auth.xboxlive.com/user/authenticate"),
  headers: _headers,
  body: body,
);
```

**Stacktrace:**

```
Unhandled exception:

#0      IOClient.send (package:http/src/io_client.dart:66:7)
<asynchronous suspension>
#1      BaseClient._sendUnstreamed (package:http/src/base_client.dart:93:32)
<asynchronous suspension>
```

**Environment:**

```
$ dart --version             
Dart SDK version: 2.15.0-116.0.dev (dev) (Thu Sep 16 09:47:01 2021 -0700) on "linux_x64"
```

**Comment by brianquinlan** [comment]
I can't be positive without a decrypted session but @Cubxity seems to have hit the same issue.

**Comment by brianquinlan** [comment]
BTW, I submitted a change 98634512f240a7e4db89f44df7dd488b684a548f that will make this sorts of errors more clear in Dart 2.70

**Comment by brianquinlan** [comment]
Fixed in c286b76c2db394b72bd8ae79b32d024c2d25c52b

**Comment by larssn** [comment]
Does this fix the original issue in https://github.com/flutter/flutter/issues/58676?
Or is the workaround simply to use `allowLegacyUnsafeRenegotiation`?
